### ISSUE TITLE
Change count_variant_alleles to not reconstruct DataArray

### ISSUE THREAD

**Issue by eric-czech** [pystatgen/sgkit#305]
I noticed that `count_variant_alleles` evaluated immediately when running it on a client to a dask cluster.  That may be because numba isn't installed on the workers.  If that was it, we should document this in a "troubleshooting" section of the docs.

**Comment by aktech** [comment]
> That may be because numba isn't installed on the workers. 

If numba isn't installed on the workers, wouldn't that have been a module not found error?

As `count_alleles`  is decorated with `guvectorize`:
https://github.com/pystatgen/sgkit/blob/037f068fc400a4aa4c124503df6f06dc918f03f4/sgkit/stats/aggregation.py#L18-L35

**Comment by eric-czech** [comment]
> If numba isn't installed on the workers, wouldn't that have been a module not found error?

Maybe?  It was nice not to have to install sgkit on the workers (yet at least) and afaik numba isn't in the dask docker image used by the default Helm chart configuration.  I'm assuming it wasn't installed and that dask might have some kind of import guard in its special logic for patching numba functions into graphs.  No idea why that would still work but not be lazy though -- will have to experiment some more.

**Comment by eric-czech** [comment]
It turns out that the problem with this has nothing to do with the cluster, but rather that a Dataset is being created incorrectly:

https://github.com/pystatgen/sgkit/blob/8467a4d4cab88747a3ec6ef8807b8c24a0e01297/sgkit/stats/aggregation.py#L209-L213

The `count_call_alleles(ds, call_genotype=call_genotype)[variables.call_allele_count].sum(dim="samples")` expression is already a DataArray.  When a DataArray is given with new dimension names, apparently that triggers an evaluation.  Here is a simple example showing this:

```python
import dask.array as da
import xarray as xr
x = da.random.randint(1, 10, size=(100, 25))
ds = xr.Dataset(dict(a=xr.DataArray(x, dims=('x', 'y'))))
type(ds.a.data)
dask.array.core.Array

# Recreate the dataset with the same array, but also redefine the dimensions
ds2 = xr.Dataset(dict(a=(('x', 'y'), ds.a))
type(ds2.a.data)
numpy.ndarray
```

**tl;dr** `xr.Dataset(dict(var=(('x', 'y'), arr)))` will never be lazy if `arr` is already `DataArray`.  We should prefer `xr.Dataset(dict(var=arr))` instead.
cc: @tomwhite, @timothymillar

**Comment by tomwhite** [comment]
Interesting. Thanks for digging into this @eric-czech.

I wonder if this is an Xarray bug (or improvement) - should it be a no-op if the new dimension names are the same as the old? Regardless, we can change our code to avoid specifying the dimensions if they are already present, and we could even add an assertion to check the dimensions are as we expect.

It would be nice if it were possible to have a mode where Dask computations log a warning message to the console. We might also think about incorporating checks that variables are lazy in our tests.

**Comment by eric-czech** [comment]
FYI: https://github.com/pydata/xarray/issues/4529

**Comment by eric-czech** [comment]
> It would be nice if it were possible to have a mode where Dask computations log a warning message to the console

`ProgressBar().register()` has been fairly useful for that with a non-distributed scheduler IMO.

> We might also think about incorporating checks that variables are lazy in our tests.

+1

**Comment by timothymillar** [comment]
Good, catch @eric-czech, +1 for testing for laziness
